Streaming support for prometheus stats

rulex123 opened the pull request:
Commit Message: patch to add support for streaming/chunking of prometheus stats - basically adapts the algorithm introduced in https://github.com/envoyproxy/envoy/pull/19693 for prometheus stats (while keeping common parts of the algorithm in a shared class to avoid code duplication). With this patch we should be able to improve the issue described here https://github.com/envoyproxy/envoy/issues/16139 

Risk Level: only underlying implementation changes, but otherwise no change of behaviour/features. So low risk I would say.
Testing: current tests should remain unchanged (though they might be moved around due to refactoring of code); new tests are also added
Docs Changes: n/a
Release Notes: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue]: https://github.com/envoyproxy/envoy/issues/16139

Preliminary benchmark data:

BEFORE
```
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
BM_AllCountersText                      668 ms          666 ms            1
BM_UsedCountersText                    40.6 ms         40.4 ms           18
BM_FilteredCountersText                6705 ms         6675 ms            1
BM_Re2FilteredCountersText              292 ms          291 ms            2
BM_AllCountersJson                      730 ms          725 ms            1
BM_UsedCountersJson                    40.4 ms         40.3 ms           17
BM_FilteredCountersJson                6822 ms         6753 ms            1
BM_Re2FilteredCountersJson              311 ms          310 ms            2
BM_AllCountersPrometheus               5489 ms         5381 ms            1
BM_UsedCountersPrometheus               160 ms          157 ms            5
BM_FilteredCountersPrometheus          7684 ms         7637 ms            1
BM_Re2FilteredCountersPrometheus        577 ms          574 ms            1
```

AFTER
```
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
BM_AllCountersText                      682 ms          680 ms            1
BM_UsedCountersText                    40.4 ms         40.3 ms           17
BM_FilteredCountersText                6749 ms         6715 ms            1
BM_Re2FilteredCountersText              297 ms          296 ms            2
BM_AllCountersJson                      715 ms          713 ms            1
BM_UsedCountersJson                    46.2 ms         45.6 ms           16
BM_FilteredCountersJson                7345 ms         7214 ms            1
BM_Re2FilteredCountersJson              323 ms          319 ms            2
BM_AllCountersPrometheus               1917 ms         1907 ms            1
BM_UsedCountersPrometheus              55.1 ms         54.6 ms           13
BM_FilteredCountersPrometheus          6886 ms         6824 ms            1
BM_Re2FilteredCountersPrometheus        305 ms          304 ms            2
```

@jmarantz for now a draft PR: I would like to get some preliminary feedback before going ahead with more refactoring (e.g. cleaning up code and moving tests around). Also it might be a good idea to loop in Greg at this point, especially to confirm I didn't overlook anything major around prometheus stats and scopes :)

jmarantz commented:
Fantastic! I will take a look!

And thanks for including the benchmarks. I am happy to review prior to
refractoring once I get back to a big screen.

On Sat, Dec 31, 2022, 10:25 AM repokitteh-read-only[bot] <
***@***.***> wrote:

> As a reminder, PRs marked as draft will not be automatically assigned
> reviewers,
> or be handled by maintainer-oncall triage.
>
> Please mark your PR as ready when you want it to be reviewed!
> 🐱
>
> Caused by: #24716 <https://github.com/envoyproxy/envoy/pull/24716> was
> opened by rulex123.
>
> see: more <https://github.com/envoyproxy/envoy/pull/24716>, trace
> <https://prod.repokitteh.app/traces/ui/envoyproxy/envoy/0a01f300-88ed-11ed-9d5d-a8ded29505d3>
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/24716#issuecomment-1368190430>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO2IPIZ2736TRTVDZK4OATWP73YZANCNFSM6AAAAAATNORLJE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jmarantz commented on source/server/admin/prometheus_stats_request.cc:
```diff
@@ -0,0 +1,160 @@
+
+    // capture stat by either adding to a pre-existing variant or by creating a new variant, and
+    // then storing the variant into the map
+    const Stats::SymbolTable& global_symbol_table = stat->constSymbolTable();
```
nit: you can save this per-stat lookup by saving the symbol table as a class member variable, since you are already capturing `[this]`

jmarantz commented on source/server/admin/prometheus_stats_request.cc:
```diff
@@ -0,0 +1,160 @@
+    const Stats::SymbolTable& global_symbol_table = stat->constSymbolTable();
+    std::string tag_extracted_name = global_symbol_table.toString(stat->tagExtractedStatName());
+
+    StatOrScopes variant = stat_map_.contains(tag_extracted_name)
```
you can make this a little faster (just one hash lookup rather than 3 in some cases) with:

```
StatOrScopes& variant = stat_map_[tag_extracted_name];
if (variant.index() == absl::variant_npos) {
  variant = std::vector<Stats::RefcountPtr<StatType>>({stat});
} else {
  absl::get<std::vector<Stats::RefcountPtr<StatType>>>(variant).emplace_back(stat);
}
```

jmarantz commented on source/server/admin/prometheus_stats_request.cc:
```diff
@@ -0,0 +1,160 @@
+  phase_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> PrometheusStatsRequest::checkStat() {
```
rename this `saveMatchingStat`? I see the original code used a local lambda `check_stat` for this, but that was mis-named as it's doing more than just checking. Might as well fix it now.

jmarantz commented on source/server/admin/prometheus_stats_request.cc:
```diff
@@ -0,0 +1,160 @@
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
```
I'm wondering if filters are ever used with Prom. It seems weird now that I think of it to filter on the fully elaborated name, but I guess that's easier from a coding and usage perspective than filtering on the tag-extracted name and then having a separate mechanism for filtering on the extracted tags.

jmarantz reviewed (commented):
I did an initial pass; let me know what you think so far.

I'm not clear on what's missing from a refactoring perspective. Is there something else we should share between prom and non-prom that you haven't done yet?

jmarantz commented:
I'm happy to take another pass when you are ready.

/wait

rulex123 commented:
Ty @jmarantz for the initial review! 

I was on leave for a few days so I didn't get a chance to look at your feedback yet, but I am back now. The refactoring to share code between the prom and non-prom stats requests classes is complete (bar the feedback you provided, which I still need to address).

The leftover refactoring concerns tests and the "old" prometheus stats code (to be removed, since it's now replaced by the `PrometheusStatsRequest` class). Hopefully I will be able to have a full-blown PR ready for you before end of week :-)

rulex123 commented on source/server/admin/base_stats_request.cc:
```diff
@@ -0,0 +1,180 @@
+namespace Server {
+
+template <class TR, class C, class G, class H>
+StatsRequestBase<TR, C, G, H>::StatsRequestBase(Stats::Store& stats, const StatsParams& params,
```
This makes sense, I will do as you suggest 👍

rulex123 commented on source/server/admin/base_stats_request.h:
```diff
@@ -0,0 +1,147 @@
+
+  // phase-related state
+  uint64_t phase_stat_count_{0};
+  unsigned short phase_;
```
The index doesn't have to do with skipping the text-readouts per se. It's used to keep track of the current phase in the shared base class, and move on to the next phase w/o having to know what that phase is.

This is because the prom and non-prom algorithms go through phases in a different order. I hope that makes sense: if you can think of a better way to achieve something similar, I am happy to update the code. For now I will rename the field name and type.

rulex123 commented:
I addressed the feedback provided so far and cleaned out the old prometheus stats code. I think the only outstanding item is to adapt pre-existing tests in `prometheus_stats_test.cc` (for now they are commented out).

RyanTheOptimist commented:
I was added as a deps reviewer but I don't see any deps changes. I'm guessing that the deps changes were removed in a previous commit. Ditto for mobile. I'll remove the labels.

jmarantz reviewed (commented):
flushing comments for now; will finish review tomorrow.

rulex123 commented on source/server/admin/stats_request.cc:
```diff
@@ -115,105 +106,76 @@ bool StatsRequest::nextChunk(Buffer::Instance& response) {
   // in the scopes.
   for (const Stats::ConstScopeSharedPtr& scope : scopes_) {
     StatOrScopes& variant = stat_map_[stats_.symbolTable().toString(scope->prefix())];
-    if (variant.index() == absl::variant_npos) {
```
It's the operator[] of the btree_map that runs a try_emplace behind the scenes and so inserts the variant into the map (and instantiates it using the variant's default constructor) when the lookup key is not in the map.

So for this reason the condition of the if statement is never true and the whole if block is redundant.

jmarantz replied:
cool thank you! Can you add that as a code-comment here in place of the code?

rulex123 commented on source/server/admin/stats_request.cc:
```diff
@@ -41,14 +35,13 @@ Http::Code StatsRequest::start(Http::ResponseHeaderMap& response_headers) {
     html_render->tableEnd(response_);
     html_render->startPre(response_);
-    render_ = std::move(html_render);
+    render_.reset(html_render.release());
```
hmm not sure what I did here: I will revert to original.

rulex123 commented on source/server/admin/stats_render.cc:
```diff
@@ -262,5 +265,66 @@ void StatsJsonRender::collectBuckets(const std::string& name,
+  const std::string tags = PrometheusStatsFormatter::formattedTags(histogram.tags());
+  const std::string hist_tags = histogram.tags().empty() ? EMPTY_STRING : (tags + ",");
+
+  const Stats::HistogramStatistics& stats = histogram.cumulativeStatistics();
```
sure thing!

rulex123 commented:
@jmarantz I've added new tests and adapted pre-existing ones now, and I also addressed any feedback you left so far - so I think this patch is finally "ready for review". 

Before I mark it as such though: I mistakenly pushed a couple of unsigned commits, is it ok if I fix by force-pushing to the branch? Not sure how to address otherwise.

jmarantz commented:
RE force-pushing -- OK -- if you could scan over the PR for unresolved comments first that would be great, otherwise they will disappear when you force-push.

Another thing I sometimes do is abandon one PR and create a new one; that way you can still reference the old PR and its comments :) But it's up to you.

jmarantz commented:
@kyessenov FYI

rulex123 commented:
Closing as this is now superseded by https://github.com/envoyproxy/envoy/pull/24998
